### PR TITLE
Resolve all PipelineResources first before continuing

### PR DESCRIPTION
**Pull request by bobcatfish**
# Changes

## Pre #1324 

As part of #1184 I need to call `GetSetup` on all PipelineResources
early on in PipelineRun execution. Since PipelineRuns declare all their
resource up front, I wanted to be able to resolve all of them at once,
then call `GetSetup` on all of them. Also, as Pipelines got more complex
(we added Conditions) it turned out we were retrieving the resources in
a few different places. Also in #1324 @pritidesai is making it so that
these Resources can be provided by spec. By resolving all of this up
front at once, we can simplify the logic later on. And you can see in
this commit that we are able to reduce the responsibilities of
ResolvePipelineRun a bit too!

## Post #1324 

In #1324 we updated PipelineRuns to allow for embedding ResourceSpecs in
PipelineRuns. This commit makes it so that the logic for resolving (i.e.
deciding if PipelineResources are specified by Spec or Ref) is shared by
PipelineRuns + TaskRuns. This is done by making it so that the "binding"
uses the same type underneath. The only reason they can't be the exact
same type is that TaskRuns additionally need the "path" attribute, which
is actually only used for PVC copying, which will be removed in #1284,
and then we should be able to remove paths entirely and the type can be
the same.

Also added some additional comments around the use of `SelfLink`, and
made sure it was well covered in the reconciler test.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [n/a] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```
For folks importing the client libraries, when ResourceBindings are instantiated for TaskRuns, they now need to explicitly declare the newly embedded struct PipelineResourceBinding.
```

### CODE REVIEW

**Comment by bobcatfish** [comment]
@pritidesai if this causes any conflicts with #1324 I'm happy to deal with the conflicts! But also maybe #1324 can get merged first and then I can deal with them in this PR :D

File: pkg/reconciler/taskrun/resources/taskresourceresolution.go [hunk 1]
```diff
@@ -70,3 +71,25 @@ func ResolveTaskResources(ts *v1alpha1.TaskSpec, taskName string, kind v1alpha1.
 	}
 	return &rtr, nil
 }
+
+// getResource will return an instance of a PipelineResource to use for r, either by getting it with getter or by
+// instantiating it from the embedded spec.
+func getResource(r *v1alpha1.TaskResourceBinding, getter GetResource) (*v1alpha1.PipelineResource, error) {
+	// Check both resource ref or resource Spec are not present. Taskrun webhook should catch this in validation error.
```
Comment (sbwsg) [inline pkg/reconciler/taskrun/resources/taskresourceresolution.go]: Very minor nit (and I see that this func has just been moved from another place, so it's totally not required to fix this now) but the block of code below this is pretty self-explanatory. This comment probably isn't needed?

File: pkg/reconciler/taskrun/resources/taskresourceresolution.go [hunk 2]
```diff
@@ -70,3 +71,25 @@ func ResolveTaskResources(ts *v1alpha1.TaskSpec, taskName string, kind v1alpha1.
...
+	}
+	if r.ResourceSpec != nil {
+		return &v1alpha1.PipelineResource{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: r.Name,
+			},
+			Spec: *r.ResourceSpec,
+		}, nil
+	}
+	return nil, xerrors.New("Neither ResourseRef not ResourceSpec is defined")
```
Comment (sbwsg) [inline pkg/reconciler/taskrun/resources/taskresourceresolution.go]: `not` -> `nor`

**Comment by bobcatfish** [comment]
I think the meow plugin broke when we updated Prow :sob: This is my theory anyway. Here's a cat (Maru?) in a box to make up for it

image

**Comment by bobcatfish** [comment]
/hold

still re-doing some things post #1324

**Comment by bobcatfish** [comment]
Okay! Now rebased on #1324, PTAL @pritidesai @sbwsg @vdemeester (note @vdemeester note the potential impact on the client usage for the CLI - maybe dashboard too? - I can avoid this by instead duplicating the logic to determine if a binding is a Ref or a Spec, but once we get to #1284 we'll probably remove "path" from the TaskRun binding so it's gonna change at that point anyway. Lemme know if you feel strongly and plz tag anyone else you think would be interested.

**Comment by pritidesai** [comment]
/lgtm

This is great and works like a charm, thanks @bobcatfish for the simplification, the code is much more readable now ❤️

**Comment by vdemeester** [comment]
> note @vdemeester note the potential impact on the client usage for the CLI - maybe dashboard too? 

The API doesn't change, so this is just a matter of bump the dependency. But both an old client or new client will still work against the API, so no need to duplicate things :wink:

**Comment by bobcatfish** [comment]
/hold cancel

Thanks for the feedback all!!

**Comment by bobcatfish** [comment]
why hello stronger linting, don't mind if i do

```
I1001 14:05:29.226] pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go:1062:9: elseif: can replace 'else {if cond {}}' with 'else if cond {}' (gocritic)
I1001 14:05:29.227] 	} else {
I1001 14:05:29.227] 	       ^
I1001 14:05:29.228] pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go:1071:9: elseif: can replace 'else {if cond {}}' with 'else if cond {}' (gocritic)
I1001 14:05:29.228] 	} else {
```

**Pull request merged**